proposals/nnnn-safe-shared-contiguous-storage.md

```diff
@@ -0,0 +1,1232 @@
+
+#### Extensions to Standard Library and Foundation types
+
+```swift
+extension Array: ContiguousStorage<Self.Element> {
+  // note: this could borrow a temporary copy of the `Array`'s storage
+  var storage: Span<Element> { get }
+}
+extension ArraySlice: ContiguousStorage<Self.Element> {
+  // note: this could borrow a temporary copy of the `ArraySlice`'s storage
+  var storage: Span<Element> { get }
+}
+extension ContiguousArray: ContiguousStorage<Self.Element> {
+  var storage: Span<Element> { get }
+}
+
+extension Foundation.Data: ContiguousStorage<UInt8> {
+  var storage: Span<UInt8> { get }
+}
+
+extension String.UTF8View: ContiguousStorage<Unicode.UTF8.CodeUnit> {
+  // note: this could borrow a temporary copy of the `String`'s storage
+  var storage: Span<Unicode.UTF8.CodeUnit> { get }
+}
+extension Substring.UTF8View: ContiguousStorage<Unicode.UTF8.CodeUnit> {
+  // note: this could borrow a temporary copy of the `Substring`'s storage
+  var storage: Span<Unicode.UTF8.CodeUnit> { get }
+}
+extension Character.UTF8View: ContiguousStorage<Unicode.UTF8.CodeUnit> {
+  // note: this could borrow a temporary copy of the `Character`'s storage
+  var storage: Span<Unicode.UTF8.CodeUnit> { get }
+}
+
+extension SIMD: ContiguousStorage<Self.Scalar> {
+  var storage: Span<Self.Scalar> { get }
+}
+extension KeyValuePairs: ContiguousStorage<(Self.Key, Self.Value)> {
+  var storage: Span<(Self.Key, Self.Value)> { get }
+}
+extension CollectionOfOne: ContiguousStorage<Element> {
+  var storage: Span<Element> { get }
+}
+
+extension Slice: ContiguousStorage where Base: ContiguousStorage {
+  var storage: Span<Base.Element> { get }
+}
+```
+
+In addition to the the safe types above gaining the `storage` property, the `UnsafeBufferPointer` family of types will also gain access to a `storage` property. This enables interoperability if `Span`-taking API. While a `Span` binding created from an `UnsafeBufferPointer` exists, the memory that underlies it must not be deinitialized or deallocated.
+
+```swift
+extension UnsafeBufferPointer: ContiguousStorage<Self.Element> {
+  // note: additional preconditions apply until the end of the scope
+  var storage: Span<Element> { get }
+}
+extension UnsafeMutableBufferPointer: ContiguousStorage<Self.Element> {
+  // note: additional preconditions apply until the end of the scope
+  var storage: Span<Element> { get }
+}
+extension UnsafeRawBufferPointer: ContiguousStorage<UInt8> {
+  // note: additional preconditions apply until the end of the scope
+  var storage: Span<UInt8> { get }
+}
+extension UnsafeMutableRawBufferPointer: ContiguousStorage<UInt8> {
+  // note: additional preconditions apply until the end of the scope
+  var storage: Span<UInt8> { get }
+}
+```
+
```

Comment:
I'm not 100% sure we want to do this generically. Elsewhere, we have an argument label `unsafe` when constructing a `Span` from unsafe pointers. 

This is unsafe-to-safe adapter code and we have explicit inits for those.

Reply:
That is fair.

Reply:
Removed in favor of an example directly using a Span initializer.